grype/match/ignore.go

```diff
@@ -0,0 +1,147 @@
+// determines if the match should be ignored, by evaluating if any of the
+// provided IgnoreRules apply to the match. If any rules apply to the match, all
+// applicable rules are attached to the Match to form an IgnoredMatch.
+// ApplyIgnoreRules returns two collections: the matches that are not being
+// ignored, and the matches that are being ignored.
+func ApplyIgnoreRules(matches Matches, rules []IgnoreRule) (Matches, []IgnoredMatch) {
+	if len(rules) == 0 {
+		return matches, nil
+	}
+
+	var ignoredMatches []IgnoredMatch
+	remainingMatches := NewMatches()
+
+	for match := range matches.Enumerate() {
+		var applicableRules []IgnoreRule
+
+		for _, rule := range rules {
+			if shouldIgnore(match, rule) {
+				applicableRules = append(applicableRules, rule)
+			}
+		}
+
+		if len(applicableRules) > 0 {
+			ignoredMatches = append(ignoredMatches, IgnoredMatch{
+				Match:              match,
+				AppliedIgnoreRules: applicableRules,
+			})
+
+			continue
+		}
+
+		remainingMatches.add(match.Package.ID, match)
+	}
+
+	return remainingMatches, ignoredMatches
+}
+
+func shouldIgnore(match Match, rule IgnoreRule) bool {
+	ignoreConditions := getIgnoreConditionsForRule(rule)
+	if len(ignoreConditions) == 0 {
+		return false
+	}
+
+	for _, condition := range ignoreConditions {
+		if !condition(match) {
+			// as soon as one rule criterion doesn't apply, we know this rule doesn't apply to the Match
+			return false
+		}
+	}
+
+	// all criteria specified in the rule apply to this Match
+	return true
+}
+
+func getIgnoreConditionsForRule(rule IgnoreRule) []ignoreCondition {
+	var ignoreConditions []ignoreCondition
+
+	if v := rule.Vulnerability; v != "" {
+		ignoreConditions = append(ignoreConditions, ifVulnerabilityApplies(v))
+	}
+
+	if n := rule.Package.Name; n != "" {
+		ignoreConditions = append(ignoreConditions, ifPackageNameApplies(n))
+	}
+
+	if v := rule.Package.Version; v != "" {
+		ignoreConditions = append(ignoreConditions, ifPackageVersionApplies(v))
+	}
+
+	if l := rule.Package.Location; l != "" {
+		ignoreConditions = append(ignoreConditions, ifPackageLocationApplies(l))
+	}
+
+	return ignoreConditions
+}
+
+type ignoreCondition func(match Match) bool
+
+func ifVulnerabilityApplies(vulnerability string) ignoreCondition {
+	return func(match Match) bool {
+		return vulnerability == match.Vulnerability.ID
+	}
+}
+
+func ifPackageNameApplies(name string) ignoreCondition {
+	return func(match Match) bool {
+		return name == match.Package.Name
+	}
+}
+
+func ifPackageVersionApplies(version string) ignoreCondition {
+	return func(match Match) bool {
+		return version == match.Package.Version
+	}
+}
+
+func ifPackageLocationApplies(location string) ignoreCondition {
+	return func(match Match) bool {
+		return locationAppliesToMatch(location, match)
+	}
+}
+
+func locationAppliesToMatch(location string, match Match) bool {
+	// TODO: This returns true if ANY of the package's locations match the rule. Is that what we want?
```

Comment:
Yes, this is probably the right behavior. 

Package locations are any evidence that leads to the existence of this package, including extra data that enriches the package info. For instance, for debian packages, we list the dpkg "status" flat file where all of the installed deb packages are listed as a "location" on the package. We also search for shared license files that may apply to the package and list those as a "location" on the package. It would be awkward if we required the location exception to require the license location to state an exception for the package as a whole.

Reply:
Agree -- I think if someone specifies some partial matches, I would expect anything matching to be excluded, the user could get more granular as needed to refine the exclusions.

Reply:
Thanks! Makes sense